### ISSUE TITLE
Limitations of basic redux pattern with generic, reusable components

### ISSUE THREAD

**Issue by bfricka** [ngrx/store#278]
I have thus far enjoyed learning ngrx and by extension, the redux pattern, but I have run in to what appears to be a very limiting flaw: It is difficult to compose generic, reusable components that hook into the store (because your payloads need to start maintaining a hierarchy of ids so actions are dispatched to the correct component, sub-component, etc).

Check out this [react/redux discussion](https://github.com/reactjs/redux/issues/822)

A couple of solutions:
- [UUIDs](https://github.com/eloytoro/react-redux-uuid)
- [multi-reducers](https://github.com/erikras/multireducer/)

I'm curious how people have approached this issue in ngrx/store. I guess to put this question more directly: If you were to implement a set of generic components such as [ng-bootstrap](https://github.com/ng-bootstrap/ng-bootstrap) using ngrx/store, how would you compose their state?

If you had 4 tabsets on a page and wanted to programmatically toggle the active tab for tabset 1, using only ngrx, you'd have to assign ids to each tabset and your payload would have to include that to select which tabset to affect when handling "SET_ACTIVE_TAB" or whatever. This is sort of cumbersome, and it's a lot of minutiae to handle manually.

Any discussion is appreciated.

**Comment by clembu** [comment]
I'm confused with your language.

> actions are dispatched to the correct component, sub-component, etc

Actions are not dispatched to components, they are dispatched to the reducers, which in turn modify the StateManager's `state`. The components may dispatch actions, though I usually hear it's best to have services do that, just to have better maintainability (if you want to switch state management systems, you'd have to refactor your services only and keep your API)

That said, regarding your tabsets problem, shouldn't a router setup suffice to indicate the navigation state?

**Comment by bfricka** [comment]
I am fully aware of how this pattern works, so what I meant by my confusing language is that when representing multiple components of the same type (and thus the same state interfaces), we start needing to identify one from the other.

Your answer about the tabsets either misinterprets my intent (apologies for the ambiguity) or (assuming you understood) is infeasible for some arbitrary collection of state representations. This is not about routing or representing state via routing.

Let's forget tabsets (I think you interpreted that as "tabs" rather than groups of tabs). Instead imagine you have 4 carousels on a single page. In order to tell carousel 1 to move to the next slide, you have to distinguish it from the other 3 carousels. Given they have the same actions and reducers it is infeasible to create distinct actions / reducers for each carousel. (e.g. `SET_SLIDE_ACTIVE_CAROUSEL_1`, `SET_SLIDE_ACTIVE_CAROUSEL_N`, etc., is obviously not a particularly elegant solution).

Instead you need a way of managing an arbitrary number of carousels. You might, for example, create a "carousels" state that is a collection of all carousels. Each carousel can then be identified by some id that is passed in with an `ADD_CAROUSEL` action. Then, when you want to move carousel 1 to the next slide, you'd dispatch the action and additionally include the id of the carousel. E.g.

```js
this.store$.dispatch({type: 'SET_ACTIVE_SLIDE', payload: {carouselId: 0, slideId: 2});
```

However, this is awkward because now everything needs to know about its ids. You need an id when you create a new carousel, and you need each action to attach this id so the correct carousel state is affected.

All of this is described in the links I originally attached, but hopefully this additional info helps clarify the discussion.

**Comment by asafper** [comment]
Hi, 

i'm struggling with the same issue , how did you tackle this eventually ?

**Comment by bfricka** [comment]
@asafper I haven't really tackled it well :). The best I could find was to break components down into generic reducers that are reusable functions (not part of the state), and then delegate as much to them as possible. Keeping tracking of multiple components still requires me to pass ids around.

**Comment by bfricka** [comment]
One way I think this could be made less problematic is if there were local state. For example if each component could access local and global state it could keep stuff related to its own context in local state while sharing the global store. However, this introduces its own set of problems. By binding a local state at the component level, you can no longer use services to dispatch events to these local stores. It's a tough problem. I'm currently struggling to elegantly implement forms using ngrx since they are a huge piece of state and should thus see the benefits of immutable state. But implementing it in a generic way is very difficult without introducing a lot of verbosity or ugliness or both.

**Comment by hccampos** [comment]
It would be great if ngrx included an easy way to create a "derived" Store bound to a slice of state and that could then be injected into the components using Angular's hierarchical injection system. This derived store could also automatically namespace actions so that reducers can be reused. To the components, they are just accessing a store, having no idea that they are actually "mounted" in a grander universe.

With something like this, components would be infinitely reusable, a la Elm or CycleJS, which, imho, is essential for building maintainable large applications.

**Comment by Juansasa** [comment]
I also struggling with the same issue and it seems like this is one area where Redux really need improvements.

The same issues have been covered in this [article](https://blog.javascripting.com/2016/05/21/the-problem-with-redux-and-how-to-fix-it/) and they propose very interesting ideas to solve these limitations.

**Comment by karolmie1** [comment]
@bfricka @asafper I've managed to create semi generic effect/reducer setup with generic effect and reducer factory: 

```
export const reducerFactory = <T>(prefix: string, actionTypes: any, keySelector: (key: T) => string) =>
    (state = initialState, action: Action): IPagedResourceState<T> => {
    const getAction = (x: string) => prefix + x;
    switch (action.type) {
        case getAction(actionTypes.SET_PAGE_STATE): {
```


```
    @Effect()
    public reloadUsers: Observable<Action> = this.actions
        .filter((action) => action.type.endsWith(baseTags.RELOAD))
        .map((x) => (x as MngActions.ReloadResoucesAction))
        .switchMap((x) =>
             this.store.select(AppReducers.MngUsersSelectors.getPageMeta).first()
                .flatMap((state) =>
                    this.resourceService.reloadUsers(x.url, state.from, state.limit))
                .map((resources) => new MngActions.SetResourcesAction(x.prefix, {resources, meta: false}))
                .catch((res) => Observable.of(new LoggerActions.LogMessageAction(message))),
        );
```

Whole repo: https://bitbucket.org/kmierzej/fragaria-webapp/src/b4b267d6582e/src/app/shared/managed-resource/?at=projects-view

I must say that it is real pain in the nuts to do this.

**Comment by bfricka** [comment]
Yeah it looks like a pain. Even looking at prism, it's not exactly a nice solution.

**Comment by dguisinger** [comment]
I have to say I'm struggling with the same problem; maybe its just that most examples of redux are terrible and overly simplified.  Very few wrap a web API, and even those use the particular data they get in only a single spot of the application, not at all making it clear what pattern to use.  I've been getting close to ripping ngrx out of my project.  I already need a service to go out and talk to the web service, and the models.  I'm starting to look at the actions, effects, and reducers and ask whats the real point of all this. 

I think similar to your example above, if you call HERO_GET, it goes out and grabs a hero from the DB and triggers HERO_GET_SUCCESS which sets the hero in the reducer and triggers an observable to fire off...  but the problem it appears you have, which I also have is, if you have more than one component requesting different Hero's.  With all of the boilerplate spaghetti code, I might as well just call my original RESTful service directly from the component and get the data.  After several weeks of trying different patterns with Redux and being stuck going in circles,  I don't get why its popular.

**Comment by nickiewooster** [comment]
[http://antontelesh.github.io/architecture/2016/03/16/fractal-architecture.html](url)

A solution for making the store composable is proposed here. I think this is quite reasonable

**Comment by PaulAGH** [comment]
I also struggled with this a great deal.  I started out implementing something along the lines of what karolmie1 in an effort to write code that could reusable, but it just got to be too much (indeed "a real pain").  I resorted to writing scripts for generating the boilerplate.  It would be ideal if ngrx allowed for a way to encapsulate all of the ngrx boilerplate into generic classes to handle all of the common logic, and then extend those classes as needed.
